src/backend/distributed/commands/call.c

```diff
@@ -0,0 +1,193 @@
+	{
+		ereport(DEBUG2, (errmsg("cannot push down CALL in multi-statement transaction")));
+		return false;
+	}
+
+	colocatedRelationId = ColocatedTableId(procedure->colocationId);
+	if (colocatedRelationId == InvalidOid)
+	{
+		ereport(DEBUG2, (errmsg("stored procedure does not have co-located tables")));
+		return false;
+	}
+
+	if (procedure->distributionArgIndex < 0 ||
+		procedure->distributionArgIndex >= list_length(funcExpr->args))
+	{
+		ereport(DEBUG2, (errmsg("cannot push down invalid distribution_argument_index")));
+		return false;
+	}
+
+	if (contain_volatile_functions((Node *) funcExpr->args))
+	{
+		ereport(DEBUG2, (errmsg("arguments in a distributed stored procedure must "
+								"be constant expressions")));
+		return false;
+	}
+
+	distTable = DistributedTableCacheEntry(colocatedRelationId);
+	partitionColumn = distTable->partitionColumn;
+
+	partitionValue = (Const *) list_nth(funcExpr->args, procedure->distributionArgIndex);
+	if (!IsA(partitionValue, Const))
+	{
+		ereport(DEBUG2, (errmsg("distribution argument value must be a constant")));
+		return false;
+	}
+
+	partitionValueDatum = partitionValue->constvalue;
+
+	if (partitionValue->consttype != partitionColumn->vartype)
+	{
+		CopyCoercionData coercionData;
+
+		ConversionPathForTypes(partitionValue->consttype, partitionColumn->vartype,
+							   &coercionData);
+
+		partitionValueDatum = CoerceColumnValue(partitionValueDatum, &coercionData);
+	}
+
+	shardInterval = FindShardInterval(partitionValueDatum, distTable);
+	if (shardInterval == NULL)
+	{
+		ereport(DEBUG2, (errmsg("cannot push down call, failed to find shard interval")));
+		return false;
+	}
+
+	placementList = FinalizedShardPlacementList(shardInterval->shardId);
+	if (list_length(placementList) != 1)
+	{
+		/* punt on reference tables for now */
+		ereport(DEBUG2, (errmsg(
+							 "cannot push down CALL for reference tables or replicated distributed tables")));
+		return false;
+	}
+
+	placement = (ShardPlacement *) linitial(placementList);
+	workerNode = FindWorkerNode(placement->nodeName, placement->nodePort);
+
+	if (workerNode == NULL || !workerNode->hasMetadata || !workerNode->metadataSynced)
+	{
+		ereport(DEBUG2, (errmsg("there is no worker node with metadata")));
+		return false;
+	}
+
+	{
+		Tuplestorestate *tupleStore = tuplestore_begin_heap(true, false, work_mem);
+		TupleDesc tupleDesc = CallStmtResultDesc(callStmt);
+		TupleTableSlot *slot = MakeSingleTupleTableSlotCompat(tupleDesc,
+															  &TTSOpsMinimalTuple);
+
+		Task *task = CitusMakeNode(Task);
+		task->jobId = INVALID_JOB_ID;
+		task->taskId = 0;
+		task->taskType = DDL_TASK;
+		task->queryString = pstrdup(queryString);
+		task->replicationModel = REPLICATION_MODEL_INVALID;
+		task->dependedTaskList = NIL;
+		task->anchorShardId = placement->shardId;
+		task->relationShardList = NIL;
+		task->taskPlacementList = placementList;
+
+		ExecuteTaskListExtended(ROW_MODIFY_COMMUTATIVE, list_make1(task),
```

Comment:
I think we can effectively treat this as a read-only query because all the necessary locking (including shard metadata locks) will be done on the worker.

Reply:
Updated to use `ROW_MODIFY_NONE` as utility statements are apt to do